csrc/fusion.h

```diff
@@ -86,7 +86,7 @@ class FusionGuard {
 
 // Set the enum base to `int` so it can be safely serialized as a part of
 // serde::InputOutputAlias.
-enum class AliasType : int {
+enum class AllocationType : int {
   NoAlias,
```

Comment:
How about the following naming for AllocationType? 

```
New,  // allocate a new buffer
ReuseBuffer,  // reuse the buffer allocated for `aliased_io`
Evaluate,  // evaluated by ExpressionEvaluator
```

Reply:
This wasn't resolved though. What's your thought on this?

Reply:
Sounds reasonable when merging `PointerArithmetic` and the current `Evaluate` but some of their requirements are different. `PointerArithmetic` requires a non-null `aliased_io` so that check should be refactored away from `allocateOutput`.

Reply:
Yes. I'd skip those checks when `aliased_io` is null. Wdyt?

Reply:
I'd still like to have some check for that (unless you think that we will never violate this condition). Let me look at if we can move it to some other place where this alias is discovered.

Reply:
> some check for that

Check for what? [These checks](https://github.com/NVIDIA/Fuser/blob/ea1bff3596fa263402c376c1ca720c5a5af53878/csrc/executor.cpp#L984-L990) only make sense when `aliased_io` is not null, so 
```
if (aliased_io != nullptr) {
  do these checks;
}
```
sounds pretty reasonable to me.

Reply:
The check for `aliased_io` itself to be non-null when it is of PointerArithmetic type. The rest of the checks can be reorganized easily. 
From a quick look at the aliasing files though, it does not seem we will run into the issue of null `aliased_io` pointers in type PointerArithmetic though.

Reply:
> The check for aliased_io itself to be non-null when it is of PointerArithmetic type.

Get it! 

> From a quick look at the aliasing files though, it does not seem we will run into the issue of null aliased_io pointers in type PointerArithmetic though.

That's right. That's why I think it's pretty safe to use whether `aliased_io==nullptr` to distinguish between PointerArithmetic and Evaluate. This way, we can remove `PointerArithmetic` whose logic is identical to `Evaluate` except the alias check. Wdyt?

Reply:
Yes, that works nicely. My only concern was regarding `PointerArithmetic` somehow ending up with null `aliased_io` which is unlikely to be the case!

Reply:
From @Priya2698:

> I actually prefer NoAlias and InplaceUpdate to New and ReuseBuffer. The former set seems more explicit to read.
> Wdyt?

SGTM -- combining the other two types is more important. 

I suggested New and ReuseBuffer because they describe an allocation, e.g., allocate a new buffer for this output vs reuse the existing buffer allocated for `aliased_io`.  AllocationType::NoAlias and AllocationType::InplaceUpdate sound a bit weird to me -- what does an inplace-update allocation mean?

Reply:
That is true. We have Alias and Allocation terminology being merged now so they don't fit well together.
I don't have a strong preference so we can move to `New` and `ReuseBuffer`. I'll add comments to explain the types instead.